### ISSUE TITLE
yamllint compatibility should be a fatal error

### ISSUE THREAD

**Issue by kousu** [ansible/ansible-lint#4227]
##### Summary

If `.yamllint` disagrees with `ansible-lint`, https://github.com/ansible/ansible-lint/pull/4139 is supposed to make it quit early with return code 3:

https://github.com/ansible/ansible-lint/blob/921aa83979873a059bf9eb257c58bde710397159/src/ansiblelint/yaml_utils.py#L141-L143

https://github.com/ansible/ansible-lint/blob/921aa83979873a059bf9eb257c58bde710397159/src/ansiblelint/constants.py#L33

but it does not.

##### Issue Type

- Bug Report

##### OS / ENVIRONMENT

```console (paste below)
$ ansible-lint --version
ansible-lint 24.6.0 using ansible-core:2.17.0 ansible-compat:24.6.0 ruamel-yaml:0.18.6 ruamel-yaml-clib:0.2.8
```

- OS: macOS 14.4.1
- ansible installation method: `brew`
- ansible-lint installation method: `brew`

##### STEPS TO REPRODUCE

```yaml
# .yamllint.yml
---
extends: default
rules:
  braces:
    min-spaces-inside: 1
    max-spaces-inside: 1
  quoted-strings:
    quote-type: double
    required: only-when-needed
```

Without linting errors:

```yaml
# site.yaml
---
- name: Base setup for all nodes
  hosts: all
  gather_facts: false
  tasks:
    - name: Install
      ansible.builtin.package:
        name:
          - firefox
          - 0ad
```

Or with:

```yaml
# site.yaml
---
- name: Base setup for all nodes
  hosts: all
  gather_facts: false
  tasks:
    - name: Install
      ansible.builtin.package:
        name:
          - "firefox"
          - 0ad
```

##### Desired Behavior

With or without linting errors should both short-circuit to:

```console
$ ansible-lint; printf "\nRC=%d\n" $?
WARNING  Found incompatible custom yamllint configuration (.yamllint.yml), please either remove the file or edit it to comply with:
  - comments.min-spaces-from-content must be 1
  - comments-indentation must be false
  - braces.min-spaces-inside must be 0
  - octal-values.forbid-implicit-octal must be true
  - octal-values.forbid-explicit-octal must be true.

Read https://ansible.readthedocs.io/projects/lint/rules/yaml/ for more details regarding why we have these requirements. Fix mode will not be available.

Failed: <some message about the short-circuit>. Last profile that met the validation criteria was 'production'.

RC=3
```

##### Actual Behavior

The linter still runs, instead of being skipped as promised.

Without linting errors:

```console
$ ansible-lint; printf "\nRC=%d\n" $?
WARNING  Found incompatible custom yamllint configuration (.yamllint.yml), please either remove the file or edit it to comply with:
  - comments.min-spaces-from-content must be 1
  - comments-indentation must be false
  - braces.min-spaces-inside must be 0
  - octal-values.forbid-implicit-octal must be true
  - octal-values.forbid-explicit-octal must be true.

Read https://ansible.readthedocs.io/projects/lint/rules/yaml/ for more details regarding why we have these requirements. Fix mode will not be available.

Passed: 0 failure(s), 0 warning(s) on 3 files. Last profile that met the validation criteria was 'production'.

RC=0
```

With linting errors:

```console (paste below)
% ansible-lint; printf "\nRC=%d\n" $?  
WARNING  Found incompatible custom yamllint configuration (.yamllint.yml), please either remove the file or edit it to comply with:
  - comments.min-spaces-from-content must be 1
  - comments-indentation must be false
  - braces.min-spaces-inside must be 0
  - octal-values.forbid-implicit-octal must be true
  - octal-values.forbid-explicit-octal must be true.

Read https://ansible.readthedocs.io/projects/lint/rules/yaml/ for more details regarding why we have these requirements. Fix mode will not be available.
WARNING  Listing 1 violation(s) that are fatal
yaml[quoted-strings]: String value is redundantly quoted with double quotes
site.yml:9

Read documentation for instructions on how to ignore specific rule violations.

                 Rule Violation Summary                  
 count tag                  profile rule associated tags 
     1 yaml[quoted-strings] basic   formatting, yaml     

Failed: 1 failure(s), 0 warning(s) on 3 files. Last profile that met the validation criteria was 'min'.

RC=2
```

**Comment by kousu** [comment]
This is super weird.  There's even a test:

https://github.com/ansible/ansible-lint/blob/921aa83979873a059bf9eb257c58bde710397159/test/test_yaml_utils.py#L996-L1002

Is there some wrapper layer betweeen `yaml_utils.py` and the actual CLI that turns all uncommon return codes into 0?

**Comment by kousu** [comment]
Oh I see what happened, the error was demoted to a warning in https://github.com/ansible/ansible-lint/pull/415. But https://github.com/ansible/ansible-lint/issues/4152 only requested it apply if `warn_list: [yaml]` or `skip_list: [yaml]` are set; otherwise it should still be an error.

I didn't set either of those in my test case above. But instead, setting

```yaml
# .ansible-lint
write_list:
  - all
```

Gives

```console
$ ansible-lint 
ERROR    Found incompatible custom yamllint configuration (.yamllint.yml), please either remove the file or edit it to comply with:
  - comments.min-spaces-from-content must be 1
  - comments-indentation must be false
  - braces.min-spaces-inside must be 0
  - octal-values.forbid-implicit-octal must be true
  - octal-values.forbid-explicit-octal must be true.

Read https://ansible.readthedocs.io/projects/lint/rules/yaml/ for more details regarding why we have these requirements. Fix mode will not be available.
$ echo $?
3
```

I don't see the connection between `write_list` and yamllint, they seem totally apart, does anyone have an explanation?

**Comment by ssbarnea** [comment]
Yes, these settings are required only if you use the `--fix` option, because we can only format using these values.

That is why for other cases we only display a warning. The user might decide to either fix ansible lint config.
